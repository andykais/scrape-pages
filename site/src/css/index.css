:root {
  --primary: #78909c;
  --primary-dark: #4b636e;
  --secondary: #bcaaa4;
  --secondary-light: white;
  --flow-error: #c40;
  --font-size: 16px;
}

html,
body {
  font-size: var(--font-size);
  padding: 0;
  margin: 0;
  font-family: sans-serif;
  height: 100%;
}
body {
  display: flex;
  flex-direction: column;
}
h1,
h2,
h3,
h4,
h5 {
  margin: 0;
}
a {
  text-decoration: none;
  position: relative;
  color: white;
}
i {
  font-size: 26px;
  line-height: 20px;
}
button i {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

#container {
  display: flex;
  flex-grow: 1;
}

@import './navbar.css';
@import './user-input.css';
@import './flow-diagram.css';
