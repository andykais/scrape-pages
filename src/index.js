import PageScraper from './classes'
import sourceMapSupport from 'source-map-support'
sourceMapSupport.install()

// TODO move this into a try catch around `run`
// I shouldnt be blocking unhandledRejections here
process.on('unhandledRejection', error => {
  console.error(error)
  process.exit(1)
})

export default PageScraper
