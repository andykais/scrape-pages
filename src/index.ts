export { scrape } from './scraper'

export * from './util/errors'

export * from './util'
