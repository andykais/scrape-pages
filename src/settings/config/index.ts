export { normalizeConfig } from './normalize'

export { flattenConfig } from './flatten'

export { assertConfigType } from './runtime/settings/config/assert'
