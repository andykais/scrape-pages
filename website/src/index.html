<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="initial-scale=1.0, width=device-width">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.2.0/css/all.css" integrity="sha384-hWVjflwFxL6sNzntih27bfxkr27PmbbK/iSvJ+a4+0owXq79v+lsFkW54bOGbiDQ" crossorigin="anonymous">
  </head>
  <body>
    <div id="navbar">
      <div id="title-container">
        <h2>scrape-pages</h2>
        <h5>config editor</h5>
      </div>
      <div class="icons">
        <span id="share">
          <i class="fas fa-share-alt"></i>
        </span>
        <a href="https://github.com/andykais/scrape-pages">
          <i class="fab fa-github"></i>
        </a>
      </div>
    </div>
    <div id="container">
      <div id="user-input">
        <textarea id="config-input" name="config-input" cols="30" rows="10"></textarea>
        <button id="normalize-button">
          <span>
            Normalize Config
          </span>
        </button>
      </div>
      <div id='hide-button-container'>
        <button id="hide-button">
          <i class="fas fa-caret-left"></i>
        </button>
      </div>
      <div id="config-dependent-content">
        <div id="tree-container"></div>
        <div id="flow-error-block"></div>
      </div>
    </div>
  </body>
</html>
