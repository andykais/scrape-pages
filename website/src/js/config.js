const config = {
  input: [],
  scrape: {
    name: "mainpage",
    download: {
      increment: 0,
      initialIndex: 0,
      headerTemplates: {},
      cookieTemplates: {},
      method: "GET",
      urlTemplate: "mainpage"
    },
    parse: { expect: "html", selector: "batchID" },
    scrapeEach: [
      {
        name: "grid",
        download: {
          increment: 1,
          initialIndex: 0,
          headerTemplates: {},
          cookieTemplates: {},
          method: "GET",
          urlTemplate: "batchPage{batchID}"
        },
        scrapeNext: {
          name: 'next-batchid',
          parse: {
            selector: "batchID",
            expect: "html"
          }
        },
        scrapeEach: [
          {
            name: "tag",
            parse: { expect: "html", selector: "span.tag" },
            scrapeEach: []
          },
          {
            name: "parse-image-page",
            parse: { expect: "html", selector: "li a.images" },
            scrapeEach: [
              {
                name: "image-page",
                download: {
                  increment: 0,
                  initialIndex: 0,
                  headerTemplates: {},
                  cookieTemplates: {},
                  method: "GET",
                  urlTemplate: "imagepage{imageID}"
                },
                parse: { expect: "html", selector: "img" },
                scrapeEach: [
                  {
                    name: "image",
                    download: {
                      increment: 0,
                      initialIndex: 0,
                      headerTemplates: {},
                      cookieTemplates: {},
                      method: "GET",
                      urlTemplate: "{src}"
                    },
                    scrapeEach: []
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
};
