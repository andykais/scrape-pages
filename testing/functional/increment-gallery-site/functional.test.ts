import os from 'os'
import path from 'path'

import { expect } from 'chai'

import { nockMockFolder } from '../../setup'
import { config } from './config'
// import { copy } from '../../../src/util/object'
import expectedQueryResult from './resources/expected-query-result.json'
import { scrape } from '../../../src'

// import * as _ from 'lodash'

const resourceFolder = `${__dirname}/resources/mock-endpoints`
const resourceUrl = 'http://increment-gallery-site.com'

const options = {
  optionsEach: {
    image: {
      read: false,
      write: true
    }
  }
}
const params = {
  folder: path.resolve(os.tmpdir(), 'scrape-pages--increment-gallery-site'),
  cleanFolder: true
}
describe('increment gallery site', () => {
  describe('with instant scraper', () => {
    const { start, query } = scrape(config, options, params)

    before(async () => {
      await nockMockFolder(resourceFolder, resourceUrl)

      const { on } = await start()
      await new Promise(resolve => on('done', resolve))
    })

    it('should group each image into a separate slot, in order', () => {
      const result = query({
        scrapers: ['image'],
        groupBy: 'image'
      })
      expect(result)
        .excludingEvery(['filename', 'id'])
        .to.be.deep.equal(expectedQueryResult.map(g => g.filter(r => r.scraper === 'image')))
    })
    it('should group tags and images together that were found on the same page', () => {
      const result = query({
        scrapers: ['image', 'tag'],
        groupBy: 'image-page'
      })
      expect(result)
        .excludingEvery(['filename', 'id'])
        .to.be.deep.equal(expectedQueryResult)
    })
  })
  describe('with cached scraper', () => {
    it('should make the expected requests on first pass', async () => {
      const count = { queued: { gallery: 0, image: 0 }, complete: { gallery: 0, image: 0 } }
      await nockMockFolder(resourceFolder, resourceUrl)

      const { start, query } = scrape(config, options, params)
      const { on } = await start()
      on('image:queued', () => count.queued.image++)
      on('image:complete', () => count.complete.image++)
      on('gallery:queued', () => count.queued.gallery++)
      on('gallery:complete', () => count.complete.gallery++)
      await new Promise(resolve => on('done', resolve))
      expect(count.queued.gallery).to.be.equal(3)
      expect(count.queued.image).to.be.equal(4)
      expect(count.complete.gallery).to.be.equal(2)
      expect(count.complete.image).to.be.equal(4)

      const result = query({
        scrapers: ['image', 'tag'],
        groupBy: 'image-page'
      })
      expect(result)
        .excludingEvery(['filename', 'id'])
        .to.be.deep.equal(expectedQueryResult)
    })
    it('on second pass, it should make zero requests', async () => {
      // const count = { queued: { gallery: 0, image: 0 }, complete: { gallery: 0, image: 0 } }
      // await nockMockFolder(resourceFolder, resourceUrl)
      // const { start, query } = scrape(
      //   config,
      //   // { ...options, cache: true },
      //   _.merge(copy(options), { cache: true, optionsEach: { gallery: { cache: false } } }),
      //   { ...params, cleanFolder: false }
      // )
      // const { on } = await start()
      // on('image:queued', () => count.queued.image++)
      // on('image:complete', () => count.complete.image++)
      // on('gallery:queued', () => count.queued.gallery++)
      // on('gallery:complete', () => count.complete.gallery++)
      // await new Promise(resolve => on('done', resolve))
      // const result = query({ scrapers: ['image'], groupBy: 'image' })
      // console.log(result.map(r => r.map(r => r.downloadData)))
      // expect(count.queued.gallery).to.be.equal(3)
      // expect(count.queued.image).to.be.equal(0)
      // expect(count.complete.gallery).to.be.equal(2)
      // expect(count.complete.image).to.be.equal(4)
      // const result = query({
      //   scrapers: ['image', 'tag'],
      //   groupBy: 'image-page'
      // })
      // expect(result)
      //   .excludingEvery(['filename', 'id'])
      //   .to.be.deep.equal(expectedQueryResult)
    })
  })

  describe('with value limit', () => {
    const configWithLimit = {
      ...config,
      scrapers: {
        ...config.scrapers,
        gallery: {
          ...config.scrapers.gallery,
          parse: { ...(config.scrapers.gallery.parse as { selector: string }), limit: 1 }
        }
      }
    }
    const { start, query } = scrape(configWithLimit, options, params)

    before(async () => {
      await nockMockFolder(resourceFolder, resourceUrl)

      const { on } = await start()
      await new Promise(resolve => on('done', resolve))
    })

    it('should group each image into a separate slot, in order', () => {
      const result = query({
        scrapers: ['image'],
        groupBy: 'image'
      })
      const expected = expectedQueryResult.map(g => g.filter(r => r.scraper === 'image'))
      expect(result)
        .excludingEvery(['filename', 'id'])
        .to.be.deep.equal([expected[0], expected[2]])
    })
  })

  describe('with psuedo-random delayed scraper', () => {
    const { start, query } = scrape(config, options, params)

    before(async () => {
      await nockMockFolder(resourceFolder, resourceUrl, { randomSeed: 1 })

      const { on } = await start()
      await new Promise(resolve => on('done', resolve))
    })

    it('should keep images and tags together, in order', () => {
      const result = query({
        scrapers: ['image', 'tag'],
        groupBy: 'image-page'
      })
      expect(result)
        .excludingEvery(['filename', 'id'])
        .to.be.deep.equal(expectedQueryResult)
    })
  })
})

/**
 * I need a way to say, use the cached response from certain scrapers if hit, and always download again for
 * others
 * - useCachedResponse? default to true?
 *
 * I need a way to say reuse existing store if found, if not, remove the files
 * - cleanFolder
 */
