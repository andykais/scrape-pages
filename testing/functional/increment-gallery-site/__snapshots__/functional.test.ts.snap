// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing/functional/increment-gallery-site/functional.test.ts testing asynchronousity & ordering instant scraper 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts testing asynchronousity & ordering instant scraper 2`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "three",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts testing asynchronousity & ordering psuedo-random delayed scraper 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "three",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with cached scraper should make the expected requests on first pass 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "three",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with instant scraper should group each image into a separate slot, in order 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with instant scraper should group tags and images together that were found on the same page 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "three",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with psuedo-random delayed scraper should keep images and tags together, in order 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "one",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "two",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "three",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "four",
      "scraper": "tag",
    },
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "five",
      "scraper": "tag",
    },
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with value limit first pass does not use the value limit 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with value limit second pass should find only the images on the first gallery page 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with value limit should group each image into a separate slot, in order 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with value limit step 1 does not use the value limit 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/quick.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/fox.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;

exports[`testing/functional/increment-gallery-site/functional.test.ts with value limit step 2 should overwrite image values and only find images from the first gallery page 1`] = `
Array [
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/the.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
  Array [
    Object {
      "byteLength": "17.0",
      "complete": 1,
      "downloadData": "[\\"http://increment-gallery-site.com/image/brown.jpg\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": true,
      "id": true,
      "parsedValue": "",
      "scraper": "image",
    },
  ],
]
`;
